src/lib/Libifl/pbsD_connect.c

```diff
@@ -470,12 +349,167 @@ __pbs_connect_extend(char *server, char *extend_data)
 	 * Disable Nagle's algorithm on the TCP connection to server.
 	 * Nagle's algorithm is hurting cmd-server communication.
 	 */
-	if (pbs_connection_set_nodelay(sock) == -1) {
-		closesocket(sock);
+	if (pbs_connection_set_nodelay(sd) == -1) {
+		CLOSESOCKET(sd);
 		pbs_errno = PBSE_SYSTEM;
 		return -1;
 	}
 
+	return sd;
+}
+
+/**
+ * @brief	Get the array of connections to all servers
+ *
+ * @param	void
+ *
+ * @return	void
+ * @retval	!NULL - success
+ * @retval	NULL - error
+ *
+ * @par Side Effects:
+ *	None
+ *
+ * @par MT-safe: Yes
+ */
+void *
+get_conn_servers(void)
+{
+	svr_conn_t *conn_arr = NULL;
+
+	conn_arr = pthread_getspecific(psi_key);
+	if (conn_arr == NULL && pbs_conf.psi != NULL) {
+		int num_svrs;
+		int i;
+
+		num_svrs = get_num_servers();
+		conn_arr = calloc(num_svrs, sizeof(svr_conn_t));
+		if (conn_arr == NULL) {
+			pbs_errno = PBSE_SYSTEM;
+			return NULL;
+		}
+
+		for (i = 0; i < num_svrs; i++) {
+			strcpy(conn_arr[i].name, pbs_conf.psi[i].name);
+			conn_arr[i].port = pbs_conf.psi[i].port;
+			conn_arr[i].sd = -1;
+			conn_arr[i].secondary_sd = -1;
+			conn_arr[i].state = SVR_CONN_STATE_DOWN;
+		}
+
+		pthread_setspecific(psi_key, conn_arr);
+	}
+
+	return conn_arr;
+}
+
+
+/**
+ * @brief	Helper function for connect_to_servers to connect to a particular server
+ *
+ * @param[in]		idx - array index for the server to connect to
+ * @param[in,out]	conn_arr - array of svr_conn_t
+ * @param[in]		extend_data - any additional data relevant for connection
+ *
+ * @return	int
+ * @retval	-1 for error
+ * @retval	fd of connection
+ */
+static int
+connect_to_server(int idx, svr_conn_t *conn_arr, char *extend_data)
+{
+	int sd = conn_arr[idx].sd;
+
+	if (conn_arr[idx].state != SVR_CONN_STATE_UP || conn_arr[idx].secondary_sd < 0) {
+		if ((sd = tcp_connect(conn_arr[idx].name, conn_arr[idx].port, extend_data)) != -1) {
+			conn_arr[idx].state = SVR_CONN_STATE_UP;
+			if (conn_arr[idx].sd > 0 && conn_arr[idx].secondary_sd < 0)
+				conn_arr[idx].secondary_sd = sd;
+			else
+				conn_arr[idx].sd = sd;
+		} else
+			conn_arr[idx].state = SVR_CONN_STATE_DOWN;
+	}
+
+	return sd;
+}
+
+/**
+ * @brief	To connect to all the servers
+ *
+ * @param[in]	server_name - name of the server to connect to (NULL if not known)
+ * @param[in]	port - port of the server to connect to (considered if server_name is not NULL)
+ * @param[in]	extend_data
+ *
+ * @return int
+ * @retval >0 - success
+ * @retval -1 - error
+ */
+static int
+connect_to_servers(char *server_name, uint port, char *extend_data)
+{
+	int i;
+	int fd = -1;
+	int ret_fd = -1;
+	int num_conf_servers = get_num_servers();
+	svr_conn_t *svr_connections = get_conn_servers();
+
+	if (svr_connections == NULL)
+		return -1;
+
+	if (server_name != NULL && pbs_conf.pbs_server_name != NULL) {
+		/* Make sure that the name matches PBS_SERVER_NAME */
+		if (!is_same_host(server_name, pbs_conf.pbs_server_name)) {
```

Comment:
Are we saying that users cannot do qstat @clusterB? There could be valid cases of user wanting to connect to a cluster that is not the one defined in its local conf file in PBS_SERVER, no?

Reply:
Good catch, i think the only check we can do here then is to see if the host is reachable or not, please let me know what you think.